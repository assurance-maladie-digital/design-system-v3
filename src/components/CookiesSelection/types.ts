export type CookieTypes = 'essentials' | 'functional' | 'analytics'

export type CookiesItems = {
	[key in CookieTypes]?: Cookie[]
}

export type Preferences = {
	[key in CookieTypes]: boolean | undefined
}

export interface Cookie {
	name: string
	description?: string
	conservation: string
}
