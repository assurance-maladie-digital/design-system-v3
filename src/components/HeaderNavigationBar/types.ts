import type { RouteLocationRaw } from 'vue-router'

export interface NavigationItem {
	label: string
	href?: string
	to?: RouteLocationRaw
}
