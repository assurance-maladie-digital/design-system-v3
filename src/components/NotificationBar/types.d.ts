export interface Notification {
	id: string
	message: string
	type: 'info' | 'success' | 'warning' | 'error'
	timeout?: number
	icon?: string | null
}
