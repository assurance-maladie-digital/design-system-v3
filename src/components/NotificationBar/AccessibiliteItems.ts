import { ExpertiseLevelEnum } from './constants/ExpertiseLevelEnum'

export const AccessibiliteItemsIndeterminate = [

	{
		title: 'Catégorie 3 : Couleurs',
		subtitle: '3.2 Dans chaque page web, le contraste entre la couleur du texte et la couleur de son arrière-plan est-il suffisamment élevé?',
		items: [
			{

				precision: '3.2.1 Dans chaque page web, le texte et le texte en image sans effet de graisse d’une taille restituée inférieure à 24px vérifient-ils une de ces conditions? Le rapport de contraste entre le texte et son arrière-plan est de 4.5:1, au moins; Un mécanisme permet à l’utilisateur d’afficher le texte avec un rapport de contraste de 4.5:1, au moins.',
				link: 'https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#3.2.1',
				solution: [{
					info1: '1. Retrouver dans le document les textes et les textes en image sans effet de graisse d’une taille restituée inférieure à 24px qui pourraient poser des problèmes de contraste ;',
					info2: '2. Pour chacun de ces textes, vérifier que : Soit le rapport de contraste entre le texte et son arrière-plan est de 4.5:1, au moins; Soit un mécanisme permet à l’utilisateur d’afficher le texte avec un rapport de contraste de 4.5:1, au moins.',
					info3: '3. Si c’est le cas pour chaque texte, le test est validé',
				}],
				expertise: ExpertiseLevelEnum.DESIGN,

			},

		],
	},
	{
		title: 'Catégorie 10 : Présentation de l\'information',
		subtitle: '10.8 Pour chaque page web, les contenus cachés ont-ils vocation à être ignorés par les technologies d’assistance',
		items: [
			{

				precision: '10.8.1 Le contenu caché a vocation à être ignoré par les technologies d’assistance , il n’a pas vocation à être ignoré par les technologies d’assistance et est rendu restituable par les technologies d’assistance suite à une action de l’utilisateur réalisable au clavier ou par tout dispositif de pointage sur un élément précédent le contenu caché ou suite à un repositionnement du focus dessus',
				link: 'https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#10.8.1',
				solution: [{
					info1: '1. Retrouver les contenus cachés (éléments pourvus de l’attribut hidden ou de l’attribut WAI-ARIA aria-hidden, ou bien d’une classe ou d’un ensemble de styles CSS susceptibles de masquer le contenu)',
					info2: '2. Pour chaque contenu caché, vérifier que :'
						+ 'Soit le contenu caché a vocation à être ignoré par les technologies d’assistance (un élément statistique de visites par exemple) ;'
						+ 'Soit le contenu caché n’a pas vocation à être ignoré par les technologies d’assistance, et dans ce cas il est rendu restituable par les technologies d’assistance au moyen :'
						+ 'Soit d’une action de l’utilisateur réalisable au clavier ou par tout dispositif de pointage sur un élément précédent le contenu caché ;'
						+ 'Soit d’une fonction de programmation qui repositionne le focus sur le contenu',
					info3: '3. Si c’est le cas pour chaque contenu caché, le test est validé',
				}],
				expertise: ExpertiseLevelEnum.DEV,

			},

		],
	},

]

export const AccessibiliteItemsValidated = [
	{
		title: 'Catégorie 1 : Images',
		subtitle: '1.2 Chaque image de décoration est-elle correctement ignorée par les technologies d’assistance ?',
		items: [
			{

				precision: '1.2.4 Chaque image vectorielle (balise <svg>) de décoration, sans légende, vérifie-t-elle ces conditions ? '
					+ ' La balise <svg> possède un attribut WAI-ARIA aria-hidden="true" ;'
					+ ' La balise <svg> et ses enfants sont dépourvus d’alternative textuelle ;'
					+ ' Les balises <title> et <desc> sont absentes ou vides ;'
					+ ' La balise <svg> et ses enfants sont dépourvus d’attribut title.',
				link: 'https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#1.2.4',
				solution: [{
					info1: '1. Pour chaque image, vérifier que l’élément <svg> ne possède pas d’attributs aria-labelledby ou aria-label et qu’il :'
						+ ' Possède un attribut WAI-ARIA aria-hidden="true" ;'
						+ ' Et est dépourvu d’alternative textuelle (ainsi que ses éléments enfants) ;'
						+ ' Et ne contient pas d’éléments <title> et <desc> à moins que vides de contenu ;'
						+ ' Et est dépourvu d’attribut title (ainsi que ses éléments enfants).'
						+ ' Le nom du composant est cohérent avec l’état de la fonctionnalité ou des contenus contrôlés (par exemple pour une fonctionnalité permettant d’afficher ou de masquer une zone de contenu).',
					info2: '2. Sinon, vérifier la présence d’un composant d’interface accessible permettant d’accéder aux mêmes fonctionnalités',
					info3: '3. Si c’est le cas pour chaque image, le test est validé.',
				}],
				expertise: ExpertiseLevelEnum.DESIGN,

			},

		],
	},
	{
		title: 'Catégorie 3 : Couleurs',
		subtitle: '3.2 Dans chaque page web, le contraste entre la couleur du texte et la couleur de son arrière-plan est-il suffisamment élevé?',
		items: [
			{

				precision: '3.2.1 Dans chaque page web, le texte et le texte en image sans effet de graisse d’une taille restituée inférieure à 24px vérifient-ils une de ces conditions? Le rapport de contraste entre le texte et son arrière-plan est de 4.5:1, au moins; Un mécanisme permet à l’utilisateur d’afficher le texte avec un rapport de contraste de 4.5:1, au moins.',
				link: 'https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#3.2.1',
				solution: [{
					info1: '1. Retrouver dans le document les textes et les textes en image sans effet de graisse d’une taille restituée inférieure à 24px qui pourraient poser des problèmes de contraste ;',
					info2: '2. Pour chacun de ces textes, vérifier que : Soit le rapport de contraste entre le texte et son arrière-plan est de 4.5:1, au moins; Soit un mécanisme permet à l’utilisateur d’afficher le texte avec un rapport de contraste de 4.5:1, au moins.',
					info3: '3. Si c’est le cas pour chaque texte, le test est validé',
				}],
				expertise: ExpertiseLevelEnum.DESIGN,

			},

		],
	},
	{
		title: 'Catégorie 7 : Scripts',
		subtitle: '7.1 Chaque script est-il, si nécessaire, compatible avec les technologies d’assistance ?',
		items: [

			{
				precision: ' 7.1.3 Chaque script qui génère ou contrôle un composant d’interface vérifie-t-il ces conditions : '
					+ 'Le composant possède un nom pertinent ;'
					+ 'Le nom accessible du composant contient au moins l’intitulé visible ;'
					+ 'Le composant possède un rôle pertinent.',

				link: 'https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#7.1.3',
				solution: [{
					info1: '1. Pour chacun des composants d’interface ayant validé le test 7.1.1, vérifier que le composant d’interface possède : ',
					info2: '2. Un nom pertinent (intitulé visible) et un rôle pertinent',
					info3: '3. Si le composant d’interface possède un nom accessible, vérifier que ce nom est pertinent et contient au moins l’intitulé visible.',
				}],
				expertise: ExpertiseLevelEnum.DESIGN,

			},

		],
	},

	{
		title: 'Catégorie 8 : Eléments Obligatoires',
		subtitle: '8.9 Dans chaque page web, les balises ne doivent pas être utilisées uniquement à des fins de présentation',
		items: [

			{
				precision: ' 8.9.1 Dans chaque page web les balises (à l’exception de <div>, <span> et <table>) ne doivent pas être utilisées uniquement à des fins de présentation. Cette règle est-elle respectée',
				link: 'https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#8.9.1',
				solution: [{
					info1: '1. Retrouver dans le document l’ensemble des éléments sémantiques utilisés à des fins de présentation ;',
					info2: '2. Pour chacun de ces éléments, vérifier que :'
						+ 'L’élément est pourvu d’un attribut role=“presentation” ;'
						+ 'L’utilisation de cet élément à des fins de présentation reste justifée',
					info3: '3. Si c’est le cas, le test est validé',
				}],
				expertise: ExpertiseLevelEnum.DESIGN,

			},

		],
	},

]
