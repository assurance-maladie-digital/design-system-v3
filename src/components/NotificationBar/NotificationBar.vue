<script setup lang="ts">
	import useCustomizableOptions, { type CustomizableOptions } from '@/composables/useCustomizableOptions'
	import { useNotificationService } from '@/services/NotificationService'
	import { mdiAlertCircleOutline, mdiAlertOctagonOutline, mdiCheckCircleOutline, mdiClose, mdiInformationOutline } from '@mdi/js'
	import { computed, getCurrentInstance, ref, watch } from 'vue'
	import { useDisplay } from 'vuetify'
	import defaultOptions from './options'
	import { type Notification } from './types'

	const props = withDefaults(defineProps<CustomizableOptions & {
		closeBtnText?: string
		rounded?: 0 | 1 | 2 | 3 | 4 | 'xs' | 'sm' | true | 'lg' | 'xl' | 'pill' | 'circle' | 'shaped'
		bottom?: true | false
	}>(), {
		closeBtnText: 'Fermer',
		rounded: 4,
		bottom: false,
	})

	const options = useCustomizableOptions(defaultOptions, props)

	const display = useDisplay()

	const { notificationQueue, removeNotification } = useNotificationService()

	const instance = getCurrentInstance()

	const currentNotification = ref<Notification>()
	const isNotificationVisible = ref(false)

	const isProcessingNotifications = ref(false) // Nouvelle variable pour suivre le traitement

	const hasActionSlot = computed(() => !!instance?.slots.action)
	const isMobileVersion = computed(() => display.name.value === 'xs')
	const isTabletVersion = computed(() => display.name.value === 'sm')
	const hasLongContent = computed(() => (currentNotification.value?.message?.length ?? 0) > 50)

	const iconMapping: Record<string, string> = {
		info: mdiInformationOutline,
		success: mdiCheckCircleOutline,
		warning: mdiAlertCircleOutline,
		error: mdiAlertOctagonOutline,
	}

	const icon = computed(() => {
		return currentNotification.value ? iconMapping[currentNotification.value.type] : null
	})

	const color = computed(() => {
		if (currentNotification.value) {
			const typeColor: Record<string, string> = {
				info: 'info',
				success: '#56C271',
				warning: '#F0B323',
				error: 'error',
			}
			return typeColor[currentNotification.value.type] || 'info'
		}
		return 'info'
	})

	const contentStyle = computed(() => {
		if (currentNotification.value) {
			const isDarkText = currentNotification.value.type === 'success' || currentNotification.value.type === 'warning'
			return {
				contentColor: isDarkText ? 'grey-darken-80' : 'white',
			}
		}
		return {
			contentColor: 'white',
		}
	})

	const smallCloseBtn = computed(() => isMobileVersion.value && !hasLongContent.value && !hasActionSlot.value)

	const setNotification = (notification: Notification) => {
		currentNotification.value = { ...notification }
	}

	const processNotificationQueue = async () => {
		isProcessingNotifications.value = true

		while (notificationQueue.value.length > 0) {
			const nextNotification = notificationQueue.value[0]
			setNotification(nextNotification)
			isNotificationVisible.value = true

			let timeout = nextNotification.timeout ?? -1

			if (timeout <= 0) {
				// Attend que la notification soit fermée manuellement car pas de timeout
				await new Promise<void>((resolve) => {
					const stopWatch = watch(isNotificationVisible, (visible) => {
						if (!visible) {
							stopWatch()
							resolve()
						}
					})
				})
			}
			else {
				// Attend la fin du délai du timeout avant de fermer la notification automatiquement
				await new Promise<void>((resolve) => {
					const timeoutId = setTimeout(() => {
						handleClearNotification()
						resolve()
					}, timeout)

					const stopWatch = watch(isNotificationVisible, (visible) => {
						if (!visible) {
							clearTimeout(timeoutId)
							stopWatch()
							resolve()
						}
					})
				})
			}

			// Retire la notification de la file
			removeNotification(nextNotification.id)
		}

		isProcessingNotifications.value = false
	}

	const handleClearNotification = () => {
		isNotificationVisible.value = false
		if (currentNotification.value) {
			removeNotification(currentNotification.value.id)
			currentNotification.value = undefined
		}
	}

	const openNotification = (notification: Notification) => {
		setNotification(notification)
		isNotificationVisible.value = true
	}

	const showNextNotification = () => {
		if (notificationQueue.value.length > 0) {
			const nextNotification = notificationQueue.value[0]
			setNotification(nextNotification)
			isNotificationVisible.value = true
		}
	}

	watch(
		() => notificationQueue.value.length,
		(newLength) => {
			if (newLength > 0 && !isProcessingNotifications.value) {
				processNotificationQueue()
			}
		},
	)

	defineExpose({
		openNotification,
		handleClearNotification,
		showNextNotification,
		processNotificationQueue,
		currentNotification,
		isNotificationVisible,
		hasActionSlot,
		isMobileVersion,
		hasLongContent,
		color,
		icon,
		contentStyle,
		smallCloseBtn,
		isVertical: computed(() => hasLongContent.value && isMobileVersion.value),
	})

</script>

<template>
	<div v-if="currentNotification">
		<VSnackbar
			v-bind="options.snackbar"
			v-model="isNotificationVisible"
			role="status"
			:color="color"
			:location="props.bottom ? 'bottom' : 'top'"
			:vertical="hasLongContent"
			:multi-line="hasLongContent"
			:timeout="currentNotification?.timeout ?? -1"
			:width="isMobileVersion || isTabletVersion ? 'auto' : '960px'"
			:rounded="props.rounded"
			:class="[{ 'long-text': hasLongContent }]"
		>
			<div class="d-flex align-center ga-2">
				<VIcon
					v-if="!isMobileVersion && icon"
					v-bind="options.icon"
					:icon="icon"
					size="24"
					aria-hidden="true"
				/>
				<p :class="'text-' + contentStyle.contentColor">
					{{ currentNotification?.message }}
				</p>
			</div>

			<template #actions>
				<div
					class="d-flex ga-2"
					style="width:100%"
					:class="hasLongContent ? 'action-section-longText' : 'action-section-shortText'"
				>
					<slot name="action" />
					<VBtn
						class="notification-bar__close"
						:class="{ 'ma-0': smallCloseBtn }"
						aria-label="Fermer la notification"
						v-bind="options.btn"
						@click="handleClearNotification"
					>
						<template v-if="!smallCloseBtn">
							{{ closeBtnText }}
						</template>
						<template v-else>
							<VIcon :icon="mdiClose" />
						</template>
					</VBtn>
				</div>
			</template>
		</VSnackbar>
	</div>
</template>

<style lang="scss" scoped>
@use '@/assets/tokens.scss';

.vd-notification-content {
  display: flex;
  align-items: center;
}

.vd-notification-bar :deep(.v-snack__wrapper) {
  padding: 16px;
  min-width: 0;
  max-width: none;
}

:deep(.v-snackbar__content) {
  padding: tokens.$padding-4 !important;
}
:deep(.v-snackbar__actions) {
  margin-inline-end: 10px;
}

.vd-notification-bar.v-snackbar--vertical :deep() {
  .v-snackbar--vertical .v-snackbar__wrapper .v-snackbar__actions {
    width: 100% !important;
    align-self: auto;
  }
  .v-snack__wrapper {
    align-items: stretch;
    flex-direction: row;
  }

  .v-snack__action {
    align-self: stretch;
    align-items: stretch;
    flex-direction: column;
  }

  .v-snackbar__content {
    margin: 0;
    width: 100%;
    display: flex;
  }

  .vd-notification-content {
    flex-direction: column;
    display: flex;
  }
}

.long-text :deep(.v-snackbar__actions) {
  width: 98% !important;
}

.short-text :deep(.v-snackbar__actions) {
  width: 48% !important;
}

.action-section-longText {
  justify-content: space-around;
}

.action-section-shortText {
  justify-content: end !important;
}
</style>
