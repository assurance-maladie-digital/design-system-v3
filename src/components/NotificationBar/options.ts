type VariantType = 'text' | 'flat' | 'elevated' | 'tonal' | 'outlined' | 'plain'
// TODO rename file to config.ts

const defaultOptions = {
	snackBar: {
		timeout: -1,
	},
	icon: {
		class: 'mr-2',
	},
	btn: {
		variant: 'text' as VariantType,
	},
}

export default defaultOptions
