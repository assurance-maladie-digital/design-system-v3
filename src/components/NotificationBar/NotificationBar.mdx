import {Controls, Canvas, Meta, Source} from '@storybook/blocks';

import * as NotificationBarStories from './NotificationBar.stories';

<Meta of={NotificationBarStories} />

# NotificationBar

Le composant `NotificationBar` est utilisé pour afficher des notifications à l’utilisateur.

<Canvas of={NotificationBarStories.Success}  />

# API
<Controls of={NotificationBarStories.Success} />

# Exemple d'utilisation

<Source dark code={`
<template>
  <div class="d-flex flex-wrap align-center justify-center">
    <NotificationBar v-model="showNotification" type="success" :closeBtnText="Fermer">
      <template #default>This is a success notification</template>
    </NotificationBar>

    <VBtn v-if="!showNotification" color="primary" @click="showNotification = true" class="ma-6">Réinitialiser</VBtn>
  </div>
</template>

<script setup lang="ts">
import NotificationBar from '@/components/NotificationBar/NotificationBar.vue'
import { ref } from 'vue'

const showNotification = ref(true);
</script>
`} />
