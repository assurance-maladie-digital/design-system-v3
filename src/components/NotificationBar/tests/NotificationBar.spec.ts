import { describe, it, expect, beforeEach } from 'vitest'
import { vi } from 'vitest'
import { mount, shallowMount } from '@vue/test-utils'
import NotificationBar from '../NotificationBar.vue'
import { type Notification } from '@/components/NotificationBar/types'
import { useNotificationService } from '@/services/NotificationService'
import { nextTick, type Ref, ref } from 'vue'
import { vuetify } from '@tests/unit/setup'

vi.mock('@/services/NotificationService')
describe('NotificationBar.vue', () => {
	interface NotificationServiceMock {
		notificationQueue: Ref<Notification[]>
		addNotification: (notification: Notification) => void
		removeNotification: (id: string) => void
	}

	let notificationServiceMock: NotificationServiceMock
	beforeEach(() => {
		notificationServiceMock = {
			notificationQueue: ref([]),
			addNotification: vi.fn(),
			removeNotification: vi.fn(),
		};
		// @ts-expect-error on vi.Mock
		(useNotificationService as vi.Mock).mockReturnValue(notificationServiceMock)
	})

	it('should render notification bar', async () => {
		const notification: Notification = {
			id: '1',
			message: 'Test message',
			type: 'info',
			timeout: -1,
			icon: null,
		}
		const wrapper = shallowMount(NotificationBar, {
			props: { notification },
			global: {
				plugins: [vuetify],
			},
		})
		wrapper.vm.openNotification(notification)

		await nextTick()
		expect(wrapper.exists()).toBe(true)
		expect(wrapper.html()).toMatchSnapshot()
	})

	it('should display the correct color based on notification type', async () => {
		const notification: Notification = {
			id: '1',
			message: 'Test message',
			type: 'success',
			timeout: -1,
			icon: null,
		}
		const wrapper = mount(NotificationBar, {
			props: { notification },
			global: {
				plugins: [vuetify],
			},
		})
		wrapper.vm.openNotification(notification)

		await nextTick()
		expect(wrapper.vm.color).toBe('success')
	})

	it('should close notification when handleClearNotification is called', async () => {
		const notification: Notification = {
			id: '1',
			message: 'Test message',
			type: 'info',
			timeout: -1,
			icon: null,
		}
		const wrapper = mount(NotificationBar, {
			props: { notification },
			global: {
				plugins: [vuetify],
			},
		})
		wrapper.vm.openNotification(notification)
		wrapper.vm.handleClearNotification()

		await nextTick()
		expect(wrapper.vm.isNotificationVisible).toBe(false)
		expect(notificationServiceMock.removeNotification).toHaveBeenCalledWith(notification.id)
	})

	it('should show the next notification in the queue', async () => {
		const notification1: Notification = {
			id: '1',
			message: 'Test message 1',
			type: 'info',
			timeout: -1,
			icon: null,
		}
		const notification2: Notification = {
			id: '2',
			message: 'Test message 2',
			type: 'success',
			timeout: -1,
			icon: null,
		}
		notificationServiceMock.notificationQueue.value = [notification1, notification2]

		const wrapper = mount(NotificationBar, {
			props: { notification: notification1 },
			global: {
				plugins: [vuetify],
			},
		})

		wrapper.vm.showNextNotification()

		await nextTick()
		expect(wrapper.vm.currentNotification).toEqual(notification1)
		expect(wrapper.vm.isNotificationVisible).toBe(true)
	})

	it('should show notification based on screen size', async () => {
		const notification: Notification = {
			id: '1',
			message: 'Test message',
			type: 'info',
			timeout: -1,
			icon: null,
		}
		const wrapper = mount(NotificationBar, {
			props: { notification },
			global: {
				plugins: [vuetify],
			},
		})
		wrapper.vm.openNotification(notification)

		await nextTick()
		expect(wrapper.vm.isMobileVersion).toBe(false)
		expect(wrapper.vm.isVertical).toBe(false)
	})

	it('should compute hasActionSlot correctly', async () => {
		const notification: Notification = {
			id: '1',
			message: 'Test message',
			type: 'info',
			timeout: -1,
			icon: null,
		}
		const wrapper = mount(NotificationBar, {
			props: { notification },
			global: {
				plugins: [vuetify],
			},
			slots: {
				action: '<div>Action Slot</div>',
			},
		})
		wrapper.vm.openNotification(notification)

		await nextTick()
		expect(wrapper.vm.hasActionSlot).toBe(true)
	})

	it('should compute isMobileVersion correctly', async () => {
		const notification: Notification = {
			id: '1',
			message: 'Test message',
			type: 'info',
			timeout: -1,
			icon: null,
		}
		const wrapper = mount(NotificationBar, {
			props: { notification },
			global: {
				plugins: [vuetify],
			},
		})
		wrapper.vm.openNotification(notification)

		await nextTick()
		expect(wrapper.vm.isMobileVersion).toBe(false)
	})

	it('should compute hasLongContent correctly', async () => {
		const notification: Notification = {
			id: '1',
			message: 'Test message'.repeat(10),
			type: 'info',
			timeout: -1,
			icon: null,
		}
		const wrapper = mount(NotificationBar, {
			props: { notification },
			global: {
				plugins: [vuetify],
			},
		})
		wrapper.vm.openNotification(notification)

		await nextTick()
		expect(wrapper.vm.hasLongContent).toBe(true)
	})

	it('should clear queue', () => {
		const isNotificationVisible = ref(true)
		const emit = vi.fn()
		const removeNotification = vi.fn()
		const currentNotification = ref({ id: 1 })

		const clearQueue = () => {
			isNotificationVisible.value = false
			emit('clear-notification')
			removeNotification(currentNotification.value.id)
		}
		clearQueue()

		expect(isNotificationVisible.value).toBe(false)
		expect(emit).toHaveBeenCalledWith('clear-notification')
		expect(removeNotification).toHaveBeenCalledWith(1)
	})

	it('should reset queue', () => {
		const notificationQueue = ref([{ id: 1 }, { id: 2 }])
		const emit = vi.fn()
		const removeNotification = vi.fn()
		const currentNotification = ref({ id: 1 })

		const resetQueue = () => {
			notificationQueue.value = []
			emit('clear-notification')
			removeNotification(currentNotification.value.id)
		}
		resetQueue()

		expect(notificationQueue.value).toEqual([])
	})

	it('should call clearQueue when resetQueue is called', () => {
		const isNotificationVisible = ref(true)
		const emit = vi.fn()
		const removeNotification = vi.fn()
		const currentNotification = ref({ id: 1 })

		const clearQueue = () => {
			isNotificationVisible.value = false
			emit('clear-notification')
			removeNotification(currentNotification.value.id)
		}

		const resetQueue = () => {
			clearQueue()
		}

		resetQueue()

		expect(isNotificationVisible.value).toBe(false)
		expect(emit).toHaveBeenCalledWith('clear-notification')
		expect(removeNotification).toHaveBeenCalledWith(1)
	})

	it('should reset all queue on resetQueue', () => {
		const notificationQueue = ref([{ id: 1 }, { id: 2 }])
		const emit = vi.fn()
		const removeNotification = vi.fn()
		const currentNotification = ref({ id: 1 })

		const clearQueue = () => {
			notificationQueue.value = []
			emit('clear-notification')
			removeNotification(currentNotification.value.id)
		}

		const resetQueue = () => {
			clearQueue()
		}

		resetQueue()

		expect(notificationQueue.value).toEqual([])
	})

	it('should handle empty notification queue', async () => {
		notificationServiceMock.notificationQueue.value = []

		const wrapper = mount(NotificationBar, {
			global: {
				plugins: [vuetify],
			},
		})

		wrapper.vm.showNextNotification()

		await nextTick()
		expect(wrapper.vm.isNotificationVisible).toBe(false)
	})

	it('should handle clearQueue when no notification is visible', () => {
		const isNotificationVisible = ref(false)
		const emit = vi.fn()
		const removeNotification = vi.fn()
		const currentNotification = ref({ id: 1 })

		const clearQueue = () => {
			isNotificationVisible.value = false
			emit('clear-notification')
			removeNotification(currentNotification.value.id)
		}
		clearQueue()

		expect(isNotificationVisible.value).toBe(false)
		expect(emit).toHaveBeenCalledWith('clear-notification')
		expect(removeNotification).toHaveBeenCalledWith(1)
	})

	it('should handle empty notification queue correctly', async () => {
		notificationServiceMock.notificationQueue.value = []

		const wrapper = mount(NotificationBar, {
			global: {
				plugins: [vuetify],
			},
		})

		await wrapper.vm.processNotificationQueue()

		expect(wrapper.vm.isNotificationVisible).toBe(false)
		expect(notificationServiceMock.removeNotification).not.toHaveBeenCalled()
	})
})
