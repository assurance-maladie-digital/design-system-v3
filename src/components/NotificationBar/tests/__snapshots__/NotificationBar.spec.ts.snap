// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotificationBar.vue > should render notification bar 1`] = `
"<div data-v-867cdd3c="" notification="[object Object]">
  <v-snackbar-stub data-v-867cdd3c="" color="info" width="960px" class="" multiline="false" timer="false" timeout="-1" vertical="false" location="top" rounded="0" tile="false" variant="elevated" absolute="false" attach="false" closeonback="true" contained="false" disabled="false" modelvalue="true" zindex="2000" activatorprops="[object Object]" openonhover="false" closeoncontentclick="false" eager="false" locationstrategy="static" origin="auto" transition="v-snackbar-transition" role="status"></v-snackbar-stub>
</div>"
`;
